recipes/cpython/all/conanfile.py

```diff
@@ -0,0 +1,514 @@
+                "MDd": "MultiThreadedDebugDLL",
+            }[str(self.settings.compiler.runtime)]
+            self.output.info("Patching runtime")
+            tools.replace_in_file(os.path.join(self._source_subfolder, "PCbuild", "pyproject.props"),
+                                  "MultiThreadedDLL", runtime_library)
+            tools.replace_in_file(os.path.join(self._source_subfolder, "PCbuild", "pyproject.props"),
+                                  "MultiThreadedDebugDLL", runtime_library)
+
+    @property
+    def _solution_projects(self):
+        solution_path = os.path.join(self._source_subfolder, "PCBuild", "pcbuild.sln")
+        project_set = set(m.group(1) for m in re.finditer("\"([^\"]+)\\.vcxproj\"", open(solution_path).read()))
+        discarded = self._msvc_discarded_projects
+        projects = set(filter(lambda p: os.path.basename(p) not in discarded, project_set))
+        return projects
+
+    @property
+    def _msvc_discarded_projects(self):
+        discarded = {"python_uwp", "pythonw_uwp"}
+        if not self.options.with_bz2:
+            discarded.add("bz2")
+        if not self.options.with_sqlite3:
+            discarded.add("_sqlite3")
+        if not self.options.with_tkinter:
+            discarded.add("_tkinter")
+        if self._is_py2:
+            # Python 2 Visual Studio projects NOT to build
+            discarded = discarded.union({"bdist_wininst", "libeay", "ssleay", "sqlite3", "tcl", "tk", "tix"})
+            if not self.options.with_bsddb:
+                discarded.add("_bsddb")
+        elif self._is_py3:
+            discarded = discarded.union({"bdist_wininst", "liblzma", "openssl", "sqlite3", "xxlimited"})
+            if not self.options.with_lzma:
+                discarded.add("_lzma")
+        return discarded
+
+    @property
+    def _msvc_archs(self):
+        archs = {
+            "x86": "Win32",
+            "x86_64": "x64",
+        }
+        if tools.Version(self.version) >= "3.8":
+            archs.update({
+                "armv7": "ARM",
+                "armv8_32": "ARM",
+                "armv8": "ARM64",
+            })
+        return archs
+
+    def _msvc_build(self):
+        msbuild = MSBuild(self)
+        msbuild_properties = {
+            "IncludeExternals": "true",
+        }
+        projects = self._solution_projects
+        self.output.info("Building {} Visual Studio projects: {}".format(len(projects), projects))
+
+        upgraded = False
+        for project_i, project in enumerate(projects):
+            self.output.info(" [{}/{}] Building project '{}'...".format(project_i, len(projects), project))
+            project_file = os.path.join(self._source_subfolder, "PCBuild", project + ".vcxproj")
+            msbuild.build(project_file, upgrade_project=not upgraded, build_type="Debug" if self.settings.build_type == "Debug" else "Release",
+                          platforms=self._msvc_archs, properties=msbuild_properties)
+            upgraded = True
```

Comment:
```suggestion
        # Python has a bootstrapping process which will run executables that are linked against it's dependencies
        with tools.run_environment(self):
            for project_i, project in enumerate(projects):
                self.output.info(" [{}/{}] Building project '{}'...".format(project_i, len(projects), project))
                project_file = os.path.join(self._source_subfolder, "PCBuild", project + ".vcxproj")
                msbuild.build(project_file, upgrade_project=not upgraded, build_type="Debug" if self.settings.build_type == "Debug" else "Release",
                                platforms=self._msvc_archs, properties=msbuild_properties)
                upgraded = True
```

Fixes building when, for example, zlib is built as a shared library.

Reply:
I've fixed this in another way by patching the vcxproj build scripts to not run python after building it.

Reply:
By the way, I had to run the following command to build with shared dependencies:
```
conan create . cpython/3.8.6@ -o cpython:shared=True -o *:shared=True -o libffi:shared=False -o mpdecimal:shared=False
```

Building cpython with shared libffi and shared mpdecimal causes compiler error.
- a shared libffi causes C2099 errors
- a shared mpdecimal causes missing symbol linking errors

Reply:
Yes, I see the same issues and I also see the CI failure for debug builds.

For the libffi error, the issue arises because static storage variables cannot be initialized with imported symbols. The best idea I've had here is to initialize the fields inside of `_ctypes_get_fielddesc`, like the unicode ctype. But this is a fairly intrusive patch, which I'm not super happy about.

mpdecimal has a few issues:

1. The fundamental source of the issue is the fact that the recipe is still using cpython's fork of the mpdecimal headers. They need to be patched out of the include path in line 101 of `_decimal.vcxproj`.
2. After you get past that, mpdecimal ships its own copy of `stdint.h` for msvc titled `vcstdint.h`, this leads to redefinition errors. To me, this is more a bug with the mpdecimal package not properly handling modern versions of msvc. For reference, cpython's fork of mpdecimal just doesn't include `vcstdint.h` in its copy of `vccompat.h`.
3. Lastly, mpdecimal has some handwritten assembly, and cpython's fork is still being included in the _decimal module. The item group starting on line 140 of `_decimal.vcxproj` should be patched out.

Not really sure what is up with the debug builds.

Reply:
Great analysis!
I will look into this coming weekend.